pkg/controller/elasticsearch/pdb/reconcile_with_roles.go

```
@@ -0,0 +1,443 @@
+			// StatefulSets with no roles are coordinating nodes - group them together
+			rolesToIndices[esv1.CoordinatingRole] = append(rolesToIndices[esv1.CoordinatingRole], i)
+			indicesToRoles[i] = set.Make(string(esv1.CoordinatingRole))
+			continue
+		}
+		for _, role := range roles {
+			// Ensure that the data* roles are grouped together.
+			normalizedRole := toGenericDataRole(role)
+			rolesToIndices[normalizedRole] = append(rolesToIndices[normalizedRole], i)
+			if _, ok := indicesToRoles[i]; !ok {
+				indicesToRoles[i] = set.Make()
+			}
+			indicesToRoles[i].Add(string(normalizedRole))
+		}
+	}
+
+	// This keeps track of which roles have been assigned to a PDB to avoid assigning the same role to multiple PDBs.
+	roleToTargetPDB := map[esv1.NodeRole]esv1.NodeRole{}
+	grouped := map[esv1.NodeRole][]int{}
+	visited := make([]bool, n)
+	for _, role := range priority {
+		indices, ok := rolesToIndices[role]
+		if !ok {
+			continue
+		}
+		for _, idx := range indices {
+			if visited[idx] {
+				continue
+			}
+			targetPDBRole := role
+			// if we already assigned a PDB for this role, use that instead
+			if target, ok := roleToTargetPDB[role]; ok {
+				targetPDBRole = target
+			}
+			grouped[targetPDBRole] = append(grouped[targetPDBRole], idx)
+			for _, r := range indicesToRoles[idx].AsSlice() {
+				roleToTargetPDB[esv1.NodeRole(r)] = targetPDBRole
+			}
+			visited[idx] = true
+		}
+	}
+	// transform into the expected format
+	res := make(map[esv1.NodeRole][]appsv1.StatefulSet)
+	for role, indices := range grouped {
+		group := make([]appsv1.StatefulSet, 0, len(indices))
+		for _, idx := range indices {
+			group = append(group, statefulSets[idx])
+		}
+		res[role] = group
+	}
+	return res
+}
+
+// getRolesFromStatefulSet extracts the roles from a StatefulSet's labels.
+func getRolesFromStatefulSet(statefulSet appsv1.StatefulSet) []esv1.NodeRole {
+	roles := []esv1.NodeRole{}
+	labels := statefulSet.Spec.Template.Labels
+	if labels == nil {
+		return roles
```

Should we return an error?